react-native-mcu-manager/android/src/main/java/uk/co/playerdata/reactnativemcumanager/DeviceUpgrade.kt

```
@@ -71,17 +79,31 @@ class DeviceUpgrade(
         val settings = Settings.Builder().setEstimatedSwapTime(estimatedSwapTime).build()
 
         try {
-            val stream = context.contentResolver.openInputStream(updateBundleUri)
-            val imageData = ByteArray(stream!!.available())
-
-            stream.read(imageData)
-
-            dfuManager.setMode(upgradeMode)
-            dfuManager.start(imageData, settings)
+            val byteArray = uriToByteArray(updateBundleUri) ?: return
+
+            try {
+                // Check if the BIN file is valid.
+                McuMgrImage.getHash(byteArray)
+
+                val stream = context.contentResolver.openInputStream(updateBundleUri)
+                val images = ByteArray(stream!!.available())
+                stream.read(images)
+                
+                dfuManager.setMode(upgradeMode)
+                dfuManager.start(images, settings)
+            } catch (e: Exception) {
+                try {
+                    val zip = ZipPackage(byteArray)
+                    val images: ImageSet = zip.getBinaries();
+                    dfuManager.setMode(upgradeMode)
+                    dfuManager.start(images, settings)
+                } catch (e1: Exception) {
+                    return
```

Again, acknowledging this is from upstream, but we're swallowing exceptions here.

We should make sure that, at the very least, the exception crashes the app.

Ideally, we should raise it to back to JavaScript land.

---

Done